### ISSUE TITLE
cfa_info handling

### ISSUE THREAD

**Issue by codido** [foss-for-synopsys-dwc-arc-processors/toolchain#122]
The closed source ARC C compiler sometimes emits the .cfa_info dwarf directive. For example, the ccac compiler emits the following for functions defined with the _Interrupt directive:
```
	.cfa_info	2	; no return pc
```

Since this isn't handled in dwarf.c, the decoding ends prematurely. As a result, objdump dumps result in something along the lines of:
```
00000050 00000018 00000000 FDE cie=00000000 pc=00000000..00000010
  DW_CFA_??? (User defined call frame op: 0x34)
```

In addition, gdb crashes during callstack unwinding if one of the functions includes the above directive:
```
/SCRATCH/arcgnu_verif/jenkins_root/ru20arcgnu2/workspace/gnu_release/gdb/gdb/dwarf2-frame.c:690: internal-error: Unknown CFI encountered.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Quit this debugging session? (y or n)
```

**Comment by claziss** [comment]
GNU tools are recognizing the CFI directives as listed [here](https://sourceware.org/binutils/docs-2.28/as/CFI-directives.html#CFI-directives). My opinion this issue is more MWDT compiler related than GNU.

**Comment by codido** [comment]
@claziss I believe this is a feature of MWDT for letting the MWDT debugger know the unwinding should stop at this point (cfa_info is a toolset-specific instruction). While it would be great to have a flag that would avoid this altogether, I couldn't find one.

However, simply disregarding it in the GNU toolchain would workaround this issue instead of crashing (the unwinding still takes one extra step, but no harm is done).

**Comment by claziss** [comment]
I've assigned Anton as this is a gdb issue.

**Comment by anthony-kolesov** [comment]
@codido  Please retry with the latest gnu toolchain release (2017.09 as of now, 2018.03-rc1 should become available soon). It seems that GDB in 2017.03 and 2018.03 doesn't crash here anymore - it just emits a warning message.

**Comment by codido** [comment]
@anthony-kolesov gdb indeed doesn't crash and emits the following instead:
`<error reading variable: Call Frame Instruction op 52 in vendor extension space is not handled on this architecture.>`

However, it's not just a warning - the frame unwinding wont be able to proceed in this case.

Also, objdump also stops at that point when dumping the dwarf frame information:
```
  DW_CFA_??? (User defined call frame op: 0x34)
```

**Comment by iguryanov** [comment]
does elfdumparc work correctly?

**Comment by codido** [comment]
elfdumparc doesn't seem to output any call frame instructions information at all, at least using the -G flag.

**Comment by iguryanov** [comment]
could you please copy here some trivial example of this function together with 2017.03 and 2017.09 objdump output, elfdumparc -G outputs, and comments on what you see and expect to have?

**Comment by anthony-kolesov** [comment]
@codido  Check if this patch makes frame unwinding work for you in the debugger: https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/commit/11a5be8f4601f536f0a121e5d55594059b3b7b2d

**Comment by codido** [comment]
@iguryanov Here's a simple example:
```
void bar();
_Interrupt void foo()
{
        bar();
}
```

objdump returns this:
```
00000050 000000a4 00000000 FDE cie=00000000 pc=00000000..00000082
  DW_CFA_??? (User defined call frame op: 0x34)
Disassembly of section .text.foo:
00000000 <foo>:
   0:   c0e1                    push_s  r0
   2:   c1e1                    push_s  r1
   4:   c2e1                    push_s  r2
   6:   c3e1                    push_s  r3
...
```

while it should return something like:
```
00000050 000000a4 00000000 FDE cie=00000000 pc=00000000..00000082
  DW_CFA_ARC_info: 2
  DW_CFA_advance_loc: 2 to 00000002
  DW_CFA_def_cfa_offset: 4
  DW_CFA_offset: r0 at cfa-4
  DW_CFA_advance_loc: 2 to 00000004
  DW_CFA_def_cfa_offset: 8
  DW_CFA_offset: r1 at cfa-8
  DW_CFA_advance_loc: 2 to 00000006
  DW_CFA_def_cfa_offset: 12
  DW_CFA_offset: r2 at cfa-12
  DW_CFA_advance_loc: 2 to 00000008
  DW_CFA_def_cfa_offset: 16
  DW_CFA_offset: r3 at cfa-16
  DW_CFA_advance_loc: 4 to 0000000c
...
```

**Comment by codido** [comment]
@anthony-kolesov I'll need to build it and test, but I suspect this wont work - the cfa info instruction has an extra argument, and seems like this would just skip over the opcode and not the argument?

**Comment by anthony-kolesov** [comment]
It doesn't seem to me that extension mechanism in GDB provides a way to parse arguments. From DWARF spec it is not clear of arguments are allowed for vendor extensions.

**Comment by iguryanov** [comment]
@codido ok, I see the effect. 
What you are looking for is output by elfdumparc -g (not -G).

arc-elf32-objdump:
0000006c 000000ac 00000000 FDE cie=00000000 pc=0000002c..000000d2
  DW_CFA_??? (User defined call frame op: 0x34)

elfdumparc:
0074: PC range .text[=2]+0x2C...text+0xD2 <===
007C: 0x34 DW_CFA_info info=2 (0x2)
007E: 0x42 DW_CFA_advance_loc +4 PC=0x4
007F: 0x0e DW_CFA_def_cfa_offset offset=4
0081: 0x9e 0x1	DW_CFA_offset reg=30 offset=-4 <===
0083: 0x41 DW_CFA_advance_loc +2 PC=0x6
0084: 0x0e DW_CFA_def_cfa_offset offset=8
0086: 0x80 0x2	DW_CFA_offset reg=0 offset=-8


where at 0x2c I have _Interrupt handler, calling external function, with prologue like this:
IRQ_Timer0:
  2c: 1cfcb788                 push	%r30 <===
  30: c0e1                     push_s	%r0
  32: c1e1                     push_s	%r1
  34: c2e1                     push_s	%r2
  36: c3e1                     push_s	%r3
....

**Comment by codido** [comment]
@anthony-kolesov Not sure about whether or not its allowed in the spec, but yeah, that requires a few more modifications probably (getting the offset in the vendor op hook and being able to advance it..).

Unfortunately, there doesn't seem to be a documented and clean way to prevent ccac from emitting these directives.

**Comment by codido** [comment]
@iguryanov By the way, while the gdb fix is a tad more tricky (since it requires changing the way vendor op hooks work), the binutils one seems to be fairly straightforward.

Something along the lines of https://gist.github.com/codido/931ea4bb489a05ac1d8c6e53c64c82c0 perhaps?

**Comment by iguryanov** [comment]
@codido maybe it will work, will check on our side. Did you already tried to built binutils this way?

**Comment by codido** [comment]
@iguryanov Yes, that seemed to work here AFAICT.

**Comment by iguryanov** [comment]
@codido ok, if it works for you can use in-house fix for now. 
I doubt that we or you could sell this change to the community. 

"0x34 DW_CFA_info info=2" is generated only by ccac to show that interrupt/exception handler does not return the same way as normal function. mcc and gcc skip this part. 

I am filing a STAR to consider different, but legit way of indicating this to mdb. 
For now please use your fixup.

**Comment by codido** [comment]
@iguryanov objdump was just given as an example for showing the incompatibility - gdb's issue is actually the more important one as this impacts its core functionality for any backtrace originating from an interrupt.

Also, there are some other platform specific changes (e.g. DW_CFA_MIPS_advance_loc8), so not sure if a fix wouldn't be upstreamable. That said, if it would be possible to just avoid this altogether by instructing the compiler not to emit such directives, that could also work.

**Comment by iguryanov** [comment]
@codido , what about DW_CFA_MIPS_advance_loc8? Does it affect you? (I had not chance before come across them). Do you have some mwdt-generated dwarf constructs which creates problems for you?

**Comment by codido** [comment]
@iguryanov No, I meant that there are platform specific CFA instructions there (e.g. DW_CFA_MIPS_advance_loc8), so might be possible to add others like DW_CFA_ARC_info.

That said, for gdb (which is more important anyway) the handler is vendor-specific anyway, so that shouldn't be a problem. It's just that the prototype isn't quite sufficient for dealing with additional arguments.

**Comment by iguryanov** [comment]
@codido ok, I see. have to check with both tool teams where it would be better to address.

**Comment by anthony-kolesov** [comment]
For GDB we can insert code into architecture independent dwarf2-frame.c - there it is possible to skip arguments. That, however, doesn't seem like to be an upstreamable, as GDB assumes that architecture extensions should be handled in architecture specific file. From the DWARF standard it is not clear if vendor extensions are allowed or forbidden to have arguments (like CCAC uses in this cfa_info). Apparently, GDB developers read standard as "arguments are forbidden". Only SPARC uses this hook in GDB, so it is hard to tell if current design is intentional or not.

**Comment by codido** [comment]
@anthony-kolesov I believe it is actually assumed that as far as the standard is concerned operands are allowed, but simply not supported in the cfa_vendor_op hook (may just be an oversight). After all, the DW_CFA_GNU_* are also extensions and they do seem to have arguments.

It seems like one possible solution would be to change the prototype of the cfa_vendor_op hook so it could be able to read operands and skip over them before the next opcode is processed.

**Comment by anthony-kolesov** [comment]
@codido  According to comments that added this hook list of arguments was decided based on what was needed by architectures that would use that hook. Description explicitly mentions that other arguments might be added if necessary.

Please check the latest arc-2018.03-gdb branch, it now would skip argument.

**Comment by codido** [comment]
@anthony-kolesov Thanks for the fix.

I haven't tried it yet, but briefly looking at the code something doesn't feel right - shouldn't you be passing a pointer to insn_ptr rather than passing it directly? Otherwise, how could it be modified by the hook?
Also, safe_skip_leb128 doesn't advance the pointer it's given in the same variable but returns the advanced pointer.

**Comment by anthony-kolesov** [comment]
Yeah, it probably will not work - new insn_ptr is returned by safe_skip_leb128, so vendor function should receive a point to insn_ptr.

Things would be much easier if you'd give me exact GDB commands that you do to reproduce the problems, so I would have an actual test case. FWIW, backtrace up to _Interrupt function is already constructed, even if argument is not skilled, so I need an actual scenario where things don't work because of the unskipped argument.

**Comment by codido** [comment]
@anthony-kolesov Unfortunately, the setup relies on internal code and tools which I'm unable to share.

However, if the simulator support for the GDB server is working properly, or you have some ARC board you can run Linux on, you can probably reproduce this by having an _Interrupt function (in code built with ccac) call another function that aborts. Alternatively, instead of an _Interrupt function, you could produce the assembly, manually add '.cfa_info 2' in that function, and assemble it.

As for the backtrace being already constructed, that may very well the be the case, but it's not printed out but rather aborted altogether, so 'bt' only shows the current frame.

**Comment by anthony-kolesov** [comment]
If frame building aborts at interrupt handler, then this handler is still visible in the backtrace. However it's local variables are not available, because frame wasn't constructed properly. I've pushed new version, it should work better. However it might produce weird backtrace behind the interrupt handler, since GDB is not aware that return address is in the ILINK, instead of BLINK.

**Comment by codido** [comment]
@anthony-kolesov Thanks!

I'll give the new patch a shot and get back to you on that.
By the way, you might want to check that insn_ptr isn't null after calling the hook, since it might be for corrupted ELF files (just in case you'd like to upstream this).

**Comment by codido** [comment]
@anthony-kolesov Just gave it a quick shot and seems to work properly, thanks!

**Comment by iguryanov** [comment]
Hi @anthony-kolesov, 

currently call frame for the handler is market this way: 
.cfa_bf foo
.cfa_info 2

What if we replace these two with ".cfa_bf foo,0"? Will it be compatible with default gdb and dwarf parsers? If so then it can be safely applied on MWDT side without upstreaming the changes in gdb.

**Comment by anthony-kolesov** [comment]
What's a .cfa_bf? Looking at the DWARF listings from previous messages it seems that this is an assembler command that doesn't turn into a specific DWARF frame operation.

**Comment by codido** [comment]
@anthony-kolesov From what I can tell, it seems to have an effect on the return register in the frame header:
```
00000000 0000004c ffffffff CIE   |  00000000 0000004c ffffffff CIE
    Version:               1     |    Version:               1
    Augmentation:          "HC"  |    Augmentation:          "HC"
    Code alignment factor: 2     |    Code alignment factor: 2
    Data alignment factor: -4    |    Data alignment factor: -4
    Return address column: 31    |    Return address column: 0
```

**Comment by iguryanov** [comment]
@anthony-kolesov, @codido 
.cfa_bf func [,linkreg] — Begin function
Directive .cfa_bf indicates the beginning of a function and must be the first CFA directive in every function. Argument func is the name of the function; the optional linkreg indicates an alternate return register (that is, if the function returns to a register other than blink). The corresponding directive .cfa_ef must be the last CFA directive in every function.

would dwarf.c/gdb complaint if return address column it some dummy r0? (by the way, can we pull return address/call stack from ERET aux? would it make sense? )

**Comment by anthony-kolesov** [comment]
If you'd move "cfa_info 2" to .cfa_bf, it still would have to generate relevant DWARF operation in frame description, so there is no difference for GDB - it reads .dwarf_frame section, not the assembly that was used to generate it :)

R0 is not dummy in ARC, unlike RISC-V, so I would think that this description can cause some confusion in the debugger, as it would try to find return address in R0. This could get into more troubles as some ARC functions might check BLINK for return address anyway, as they don't take this DWARF information into consideration.

Changing this into ERET is possible in theory, but it is complicated because this is an AUX register, therefore we'd need to translate a DWARF register number in to the AUX register number - and this should be done in a manner consistent across compilers and debugger. It is much easier with core registers, since it is implicitly assumed that DWARF register number is equal to core register number and AUX registers are not saved on the stack. It would be better to do this via debugger frame analysis, then through DWARF information description. Using ILINK would be simple, since this is a core register with clearly defined number.

**Comment by iguryanov** [comment]
we could remove confusing "cfa_info 2" completely. 
About attempt to interpret R0 - I was afraid this would be the case. 
ILINK would be simpler, but we are not choosing where to push return address :-)

**Comment by vineetgarc** [comment]
FWIW the standard way to terminate call stack is to insert a CFI which marks invalid return register.
e.g. in ARC glibc port we do this

```
       .cfi_label .Ldummy
       cfi_undefined (blink)
```

**Comment by anthony-kolesov** [comment]
CFI are assembler directives, gas converts CFI directives to DWARF CFA opcodes in either `.en_frame` or `.debug_frame` section. User/compiler can either use CFI directives or DWARF CFA opcodes to generate debug information - it is up to them, though CFI is, of course, much more convenient, because it is higher level; but the end result is always CFA opcodes. This issue is about handling of CFA opcodes generated, not about whether input code used CFA or CFI.

**Comment by vineetgarc** [comment]
I understand all that.
As I read it the original issue is related to input (and its implications on objdump/gdb etc)
"The closed source ARC C compiler sometimes emits the .cfa_info 2 dwarf directive" 

So if ccac were to stop generating the custom dwarf direcrive (assuming it does so today) then the ensuing problems would be avoided.

**Comment by abrodkin** [comment]
Fixed with https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/commit/ee4f024fa1dd257bae16b9d2dc4b1a6e3e21790b

**Comment by shahab-vahedi** [comment]
@abrodkin please don't close this issue. The patch that you referred to has been there for a long time and that is not a final fix for this issue.

**Comment by abrodkin** [comment]
@shahab-vahedi do we still want to keep this on open? If so maybe we move it to the next release `arc-2022.03`?

**Comment by shahab-vahedi** [comment]
This issue will only be closed when MW compiler complies to the standard. That means it will not be closed for this release nor the next one. If this is a point of concern, I will move it to `binutils-gdb` repo.

**Comment by abrodkin** [comment]
That's fine, I'll reset a milestone here and that's it. Let's keep it sitting here.

**Comment by shahab-vahedi** [comment]
Here is a sample test case that does produce a binary with the nefarious `cfa_info 2`. Just rename it to `*.tgz`.

[cfa_vendor.zip](https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/files/7340513/cfa_vendor.zip)

With that test case at hand, I'll try to find a solution from GDB side that can be upstreamed. If I fail at that, an internal ticket will be opened that MW team must fix this behaviour and GDB will drop these patches after a grace period of time. So far the alternatives for MW compiler [1] are:

1) Add a flag that turns off emitting the `.cfa_info x` opcodes
2) As Vineet suggested, they could generate the same opcode that derives from `cfi_undefined(x)`. Because if `glibc` uses that, there is no way that `cfa_undefined(blink)` leads to an opcode of `.cfa_info 2`. GCC just doesn't do that. There's one caveat here: This will terminate the backtracing, while in an ideal situation we would want GDB to switch to `ilink` for that particular frame's backtracing. On the other hand, GDB, as is, doesn't do that either. So at least nothing will deteriorate and things are more homogenous.

[1]
@iguryanov meticulously brought to my attention that even `mcc` (High C Compiler) generates those `.cfa_info x` opcodes. That's probably the legacy that led to all of this.

**Comment by iguryanov** [comment]
Hi, 

this was actually addressed a year ago with MWDT 2020.09  (for **ccac/arcv2**). 

Now interrupt handlers are annotated with "	.cfa_undefined %blink           "

And here is how it looks in dwarf info:

00000164 0000005c 00000100 FDE cie=00000100 pc=00000130..000001c0
  **DW_CFA_undefined: r31**
  DW_CFA_advance_loc: 4 to 00000134

016C: PC range 0x130..0x1C0
**0174: 0x07 DW_CFA_undefined reg=31**
0176: 0x42 DW_CFA_advance_loc +4 PC=0x134

I forgot to update issue, sorry.

**Comment by shahab-vahedi** [comment]
Hi Igor. I've used the engineering build of MetaWare compiler from July 2020. That's probably why I was still able to get the `cfa_info 2`. With this new piece of information you provided, there will be no effort in upstreaming anything for GDB. It's just a matter of time that the current patches will be dropped. Probably from 2022.09. That would be almost 2 years after the compiler update.
